### PR TITLE
feat(orc8r): add control_proxy.yml endpoints

### PR DESCRIPTION
**Pull request by reginawang3495**
## Summary
Add 2 endpoints:
image

## Test Plan
- Added tests
- Ran orc8r tests
- Ran through curl requests manually in Swagger UI

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by hcgatewood — COMMENTED** [review]
Shape looks great 👍

File: orc8r/cloud/go/services/tenants/servicers/servicer.go [hunk 1]
```diff
@@ -98,3 +98,38 @@ func (s *tenantsServicer) DeleteTenant(c context.Context, request *protos.GetTen
...
+}
+
+func (s *tenantsServicer) CreateOrUpdateControlProxy(c context.Context, request *protos.IDAndControlProxy) (*protos.Void, error) {
+	_, err := s.store.GetTenant(request.Id)
+	switch {
+	case err == errors.ErrNotFound:
+		return nil, status.Errorf(codes.NotFound, "Tenant %d not found", request.Id)
+	case err != nil:
+		return nil, status.Errorf(codes.Internal, "Error getting tenant %d: %v", request.Id, err)
+	}
```
Comment (reginawang3495) [inline orc8r/cloud/go/services/tenants/servicers/servicer.go]: Duplicated a couple times -- can be extracted if desired. 

**Reasoning for not doing so:** I like as is because it's easy to read and abstracting too much makes the code hard to follow sometimes

File: orc8r/protos/tenants.proto [hunk 2]
```diff
@@ -35,10 +35,17 @@ message IDAndTenant {
     Tenant tenant = 2;
```
Comment (reginawang3495) [inline orc8r/protos/tenants.proto:35]:
```
message GetTenantRequest {
    int64 id = 1;
}
```
Could be renamed to something more `controlProxy` inclusive if desired

**Reasoning for not doing so:** I don't think this is necessary since we are living in the `tenants` service anyhow

**Review by reginawang3495 — COMMENTED** [review]
2 things I don't think are necessary, but improvement ideas to point out!

File: orc8r/cloud/go/services/tenants/client_api.go [hunk 3]
```diff
@@ -126,3 +126,39 @@ func DeleteTenant(ctx context.Context, tenantID int64) error {
...
+	}
+
+	controlProxy, err := oc.GetControlProxy(ctx, &protos.GetTenantRequest{Id: tenantID})
+	if err != nil {
+		switch {
+		case status.Convert(err).Code() == codes.NotFound:
+			return controlProxy, merrors.ErrNotFound
+		default:
+			return controlProxy, err
+		}
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/client_api.go]: Let's pull this switch into a separate helper fn

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go [hunk 4]
```diff
@@ -148,3 +159,55 @@ func DeleteTenantHandler(c echo.Context) error {
...
+		return terr
+	}
+
+	_, err := tenants.GetTenant(c.Request().Context(), tenantID)
+	switch {
+	case err == errors.ErrNotFound:
+		return echo.NewHTTPError(http.StatusNotFound, fmt.Errorf("Tenant %d does not exist", tenantID))
+	case err != nil:
+		return echo.NewHTTPError(http.StatusInternalServerError, err)
+	}
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Same here -- let's pull switch into a helper fn

File: orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml [hunk 5]
```diff
@@ -94,6 +94,38 @@ paths:
         default:
           $ref: './orc8r-swagger-common.yml#/responses/UnexpectedError'
 
+  /tenants/{tenant_id}/control_proxy:
+    get:
+      summary: Retrieve ControlProxy Info By Tenant ID
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]: Here and below: sentence-case

File: orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml [hunk 6]
```diff
@@ -94,6 +94,38 @@ paths:
...
+        default:
+          $ref: './orc8r-swagger-common.yml#/responses/UnexpectedError'
+    put:
+      summary: Create or Update ControlProxy
+      tags:
+        - Tenants
+      parameters:
+        - $ref: './orc8r-swagger-common.yml#/parameters/tenant_id'
+        - in: body
+          name: controlProxy
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]: Here and elsewhere: snake_case

File: orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml [hunk 7]
```diff
@@ -114,3 +146,17 @@ definitions:
         type: array
         items:
           type: string
+  control_proxy:
+    type: object
+    required:
+      - id
+      - control_proxy
+    properties:
+      id:
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]: Unclear that we need ID in this object -- since we're only PUT-ing, we can grab the ID from the PUT endpoint

File: orc8r/protos/tenants.proto [hunk 8]
```diff
@@ -35,10 +35,17 @@ message IDAndTenant {
...
+}
+
 service TenantsService {
     rpc GetAllTenants (Void) returns (TenantList) {}
     rpc GetTenant (GetTenantRequest) returns (Tenant) {}
     rpc CreateTenant (IDAndTenant) returns (Void) {}
     rpc SetTenant (IDAndTenant) returns (Void) {}
     rpc DeleteTenant (GetTenantRequest) returns (Void) {}
+    rpc GetControlProxy (GetTenantRequest) returns (IDAndControlProxy) {}
+    rpc CreateOrUpdateControlProxy (IDAndControlProxy) returns (Void) {}
```
Comment (hcgatewood) [inline orc8r/protos/tenants.proto]: We always want the req and res to be `<rpc name>Request` and `<rpc name>Response` -- I know we don't follow that pattern correctly everywhere 🙂

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Couple small things, overall looks great though 👍

**Comment by hcgatewood** [comment]
cc @sudhikan

File: orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml (hunk 7 above)
Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]: We could, but I wanted to be same as `tenant` (see L130-148) 

Thoughts?

Comment (hcgatewood) [reply orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]: I generally prefer conciseness (e.g. don't repeat a value in both the key and value), but I also get that there's benefit to a value being able to self-identify. I'll let you make the decision then

Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml:155]: Maybe I can make an issue for removing `id` for both?

File: orc8r/protos/tenants.proto (hunk 8 above)
Comment (reginawang3495) [reply orc8r/protos/tenants.proto]: Will do, but also just wondering why? Since the message is the same thing duplicated

Comment (hcgatewood) [reply orc8r/protos/tenants.proto]: For sure. So the reason is backward compatibility -- *currently* they are the same message (same fields), but that might change over time

Specifically, to preserve proto/grpc backward compatibility, there are a specific set of rules, but basically, if you only append new fields, you're in the clear

Potential issues of sharing, especially when a large number of rpcs share 1 underlying msg type

1. RPCs could diverge in purpose over time, which would cause the shared message type to effectively multiplex multiple logical message types, which causes confusion and larger surface for bugs
2. We could need to deprecate a field, but just for 1 of the services -- no clear solution in that case, so it just wouldn't get deprecated

More info: https://github.com/uber/prototool/blob/dev/style/README.md#servicesrpcs

Comment (hcgatewood) [reply orc8r/protos/tenants.proto]: Obv this isn't really an issue for specifically our orc8r-internal grpc/proto defns, but it's good practice to keep this consistent across all gprc/proto defns

Comment (reginawang3495) [reply orc8r/protos/tenants.proto]: Thanks for all the info!

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go (hunk 4 above)
Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: I did it, but I'm not sure it's an improvement since it's provides less clarity on exactly how the error handling is + not much less code

Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: I could also theoretically pass in a map of `StatusInternalServerError` to `error`, but I don't think that really decreases the code since that's the same thing as a `switch case`

File: orc8r/cloud/go/services/tenants/servicers/storage/storage.go [hunk 9]
```diff
@@ -147,6 +150,47 @@ func (b *blobstoreStore) DeleteTenant(tenantID int64) error {
...
+		return err
+	}
+	defer store.Rollback()
+
+	controlProxyBlob := blobstore.Blob{
+		Type:  tenants.ControlProxyInfoType,
+		Key:   strconv.FormatInt(tenantID, 10),
+		Value: []byte(controlProxy),
+	}
+	glog.Infof("controlProxyBlob: %v type %v ket %v tid %v", controlProxyBlob, controlProxyBlob.Type, controlProxyBlob.Key, tenantID)
```
Comment (reginawang3495) [inline orc8r/cloud/go/services/tenants/servicers/storage/storage.go:184]: here

File: orc8r/cloud/go/services/tenants/servicers/servicer.go [hunk 10]
```diff
@@ -98,3 +95,43 @@ func (s *tenantsServicer) DeleteTenant(c context.Context, request *protos.GetTen
...
+	err = errorHandlingForGet(err, "controlProxy", request.Id)
+	if err != nil {
+		return nil, err
+	}
+	return &protos.GetControlProxyResponse{Id: request.Id, ControlProxy: controlProxy}, nil
+}
+
+func (s *tenantsServicer) CreateOrUpdateControlProxy(c context.Context, request *protos.CreateOrUpdateControlProxyRequest) (*protos.Void, error) {
+	_, err := s.store.GetTenant(request.Id)
+	glog.Infof("requestid %v", request.Id)
```
Comment (reginawang3495) [inline orc8r/cloud/go/services/tenants/servicers/servicer.go]: here

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go (hunk 4 above)
Comment (hcgatewood) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Mapping errors like this from a func is preferable imo. It 1. encapsulates the repeated logic in a single location, 2. yes it hides the impl, but you can jump to defn once and then you understand what it's doing, and 3. it strongly improves the skim-ability of the code

File: orc8r/cloud/go/services/tenants/client_api.go [hunk 11]
```diff
@@ -72,17 +72,8 @@ func GetTenant(ctx context.Context, tenantID int64) (*protos.Tenant, error) {
...
-	if err != nil {
-		switch {
-		case status.Convert(err).Code() == codes.NotFound:
-			return tenant, merrors.ErrNotFound
-		default:
-			return tenant, err
-		}
-	}
-	return tenant, nil
+	return tenant, errorHandling(err)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/client_api.go]: General: we still want to have the "if err != nil" block in this code. It's just the mapping between error types that we want to encapsulate

Also, when there's an err, we want to return zero-y values for all the non-err return vars. That's a good general default, because we don't want callers to become reliant on this implicit/unofficial part of the function contract

So it would be e.g.

```
if err != nil {
	return nil, mapErr(err)
}

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go [hunk 12]
```diff
@@ -103,11 +114,9 @@ func GetTenantHandler(c echo.Context) error {
 		return terr
 	}
 	tenantInfo, err := tenants.GetTenant(c.Request().Context(), tenantID)
-	switch {
-	case err == errors.ErrNotFound:
-		return echo.NewHTTPError(http.StatusNotFound, fmt.Errorf("Tenant %d does not exist", tenantID))
-	case err != nil:
-		return echo.NewHTTPError(http.StatusInternalServerError, err)
+	err = errorHandlingForGet(err, fmt.Errorf("Tenant %d does not exist", tenantID), err)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Same idea here -- call this `mapErr` (or `mapGetErr` if necessary), and only perform the wrap as part of the return, so e.g. `return mapErr(err)`

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go [hunk 13]
```diff
@@ -140,11 +147,65 @@ func DeleteTenantHandler(c echo.Context) error {
...
+	tenantID, terr := obsidian.GetTenantID(c)
+	if terr != nil {
+		return terr
+	}
+
+	_, err := tenants.GetTenant(c.Request().Context(), tenantID)
+	err = errorHandlingForGet(err, fmt.Errorf("Tenant %d does not exist", tenantID), err)
+	if err != nil {
+		return err
+	}
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Nit: here and below, we could refactor this shared functionality out to `validateTenantExists` function

File: orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml [hunk 14]
```diff
@@ -114,3 +146,11 @@ definitions:
         type: array
         items:
           type: string
+  control_proxy:
+    type: object
+    required:
+      - control_proxy
+    properties:
+      control_proxy:
+        description: Control proxy of the tenant
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/models/swagger.v1.yml]:
```suggestion
        description: Contents of a gateway control_proxy.yml
```

File: orc8r/cloud/go/services/tenants/servicers/servicer.go [hunk 15]
```diff
@@ -63,23 +63,19 @@ func (s *tenantsServicer) CreateTenant(c context.Context, request *protos.IDAndT
...
 func (s *tenantsServicer) GetTenant(c context.Context, request *protos.GetTenantRequest) (*protos.Tenant, error) {
 	tenant, err := s.store.GetTenant(request.Id)
-	switch {
-	case err == errors.ErrNotFound:
-		return nil, status.Errorf(codes.NotFound, "Tenant %d not found", request.Id)
-	case err != nil:
-		return nil, status.Errorf(codes.Internal, "Error getting tenant %d: %v", request.Id, err)
+	err = errorHandlingForGet(err, "tenant", request.Id)
+	if err != nil {
+		return nil, err
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/servicers/servicer.go]: Same idea here and below

File: orc8r/protos/tenants.proto [hunk 16]
```diff
@@ -35,10 +35,22 @@ message IDAndTenant {
...
+}
+
 service TenantsService {
     rpc GetAllTenants (Void) returns (TenantList) {}
     rpc GetTenant (GetTenantRequest) returns (Tenant) {}
     rpc CreateTenant (IDAndTenant) returns (Void) {}
     rpc SetTenant (IDAndTenant) returns (Void) {}
     rpc DeleteTenant (GetTenantRequest) returns (Void) {}
+    rpc GetControlProxy (GetTenantRequest) returns (GetControlProxyResponse) {}
+    rpc CreateOrUpdateControlProxy (CreateOrUpdateControlProxyRequest) returns (Void) {}
```
Comment (hcgatewood) [inline orc8r/protos/tenants.proto]: These should both be `rpc <name> (<name>Request) returns (<name>Response) {}`

**Review by hcgatewood — APPROVED** [review]
Looks great! Couple small requests on the error handling then msg me to merge 👍

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go (hunk 12 above)
Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: this is only for get though, (otherwise would need more paramters), can I call it `mapErrForGet`?

File: orc8r/protos/tenants.proto (hunk 16 above)
Comment (reginawang3495) [reply orc8r/protos/tenants.proto]: Even the void?

File: orc8r/cloud/go/services/tenants/servicers/storage/storage.go [hunk 17]
```diff
@@ -20,6 +20,8 @@ type Store interface {
 	GetAllTenants() (*protos.TenantList, error)
 	SetTenant(tenantID int64, tenant protos.Tenant) error
```
Comment (reginawang3495) [inline orc8r/cloud/go/services/tenants/servicers/storage/storage.go:21]: implementation of `SetTenant` and `CreateTenant` are exactly the same. Prefer code reuse

File: orc8r/protos/tenants.proto (hunk 16 above)
Comment (hcgatewood) [reply orc8r/protos/tenants.proto]: Correct -- cause e.g. if we suddenly needed to return something in that return proto, we'd have to add the field to `Void` to remain backwards-compatible

Again, not a big issue for our purposes, but is a good general pattern to follow

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go (hunk 12 above)
Comment (hcgatewood) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: I like `mapGetErr` -- conciseness/skim-ability is a huge pro for Go code. But `mapErrForGet` is also fine, up to you 👍

Comment (reginawang3495) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: But the method functionality could be confused for that "the method gets err"?

Comment (hcgatewood) [reply orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Up to you 🙂

File: orc8r/cloud/go/services/tenants/servicers/storage/storage_test.go [hunk 18]
```diff
@@ -161,3 +165,35 @@ func TestBlobstoreStore_DeleteTenant(t *testing.T) {
 	err = s.DeleteTenant(0)
 	assert.EqualError(t, err, "error")
 }
+
+// TODO(reginawang3495): Do or make ticket for making other tests in this file use less mocking
```
Comment (reginawang3495) [inline orc8r/cloud/go/services/tenants/servicers/storage/storage_test.go]: xx

Comment (hcgatewood) [reply orc8r/cloud/go/services/tenants/servicers/storage/storage_test.go]: This amount of testing seems sufficient 👍

File: orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go [hunk 19]
```diff
@@ -139,12 +147,75 @@ func DeleteTenantHandler(c echo.Context) error {
...
+
+	err := validateTenantExists(c, tenantID)
+	if err != nil {
+		return err
+	}
+
+	controlProxy, err := tenants.GetControlProxy(c.Request().Context(), tenantID)
+	if err != nil {
+		return mapErr(err, fmt.Errorf("control proxy for tenantID %d does not exist", tenantID), err)
+		return mapErr(err, fmt.Errorf("control proxy for tenantID %d does not exist", tenantID), err)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/tenants/obsidian/handlers/handlers.go]: Duplicate

**Review by hcgatewood — APPROVED** [review]
Lgtm! One small fix then msg to merge
